mui/mui-x
# [TimePicker] TimeClock 24 hr face (avoid ampm buttons)

# FEATURE / DESIGN DEMO (definitely do not merge (yet))

- [x] I have followed (at least) the [PR section of the contributing guide](https://github.com/mui/mui-x/blob/HEAD/CONTRIBUTING.md#sending-a-pull-request).

In Light Mode:
<img width="334" alt="SCR-20241116-kkfo" src="https://github.com/user-attachments/assets/9080c426-dec8-41c6-a494-1047c205419e">

In Dark mode:
<img width="296" alt="SCR-20241116-lhxb" src="https://github.com/user-attachments/assets/af4ebbab-719a-4ddb-9402-6a335c776a9e">


# changes

Redesign of (round) Clock component of TimeClock for:
- remove extra click requirement for am/pm (building time-tracking app and extra click really adds up, plus many users forget to choose correct am/pm after picking time)
- provide loud visual queues to user indicating the am/pm-ness of their hour number choice.
- moved 13-23 numbers on 24h clock from inner to outer ring (to match new 24-slot ampm dial) (users confused inner number for minute entry numbers and would pick wrong hour number (2 instead of 14) thinking that inner 13-23 numbers are minutes)
- Minutes dial now has notches to hint that sub-5-minute entry is possible if you just click on the notches. (Users were upset that only 12 options 00-55 were presented for selection and did not have a visual queue indicating sub-5-minute selection is possible)

![SCR-20241116-leuv](https://github.com/user-attachments/assets/793b249e-497c-4c45-b003-f14cc34848dc)


# about

This redesign was targeting specific narrow use of `DesktopTimePicker` component in compact (mobile-compatible) renderer presentation of popup:

```
<DesktopTimePicker
  label="Start Time"
  viewRenderers={{
    hours: renderTimeViewClock,
    minutes: renderTimeViewClock,
    seconds: renderTimeViewClock,
  }}
/>
```

The 3-column popup selector (hour, minute, ampm) that is default for DesktopTimePicker will be ignored in our use. We contemplate switching it to 2-click system too with values in `hour` column being a list of 24 labels `['am 1', 'am 2', ... 'pm 11']` (where internal value of each label being hour in 24h format which completely avoids a need for 3rd ampm selector) as opposed to 12 elements `['1',..,'12']` requiring additional 3rd ampm selector. However, combination of manual-entry-compatible DesktopTimePicker input field with round-clock picker was deemed to be completely compatible with all rendering envs we target (from desktop to tablets to mobile). So, we did not alter / test 3-column time selector component correspondingly to the round dial changes.

We are comfortable maintaining our own copy of 24-h round TimeClock for our uses, but are absolutely willing to help / contribute it to the project. If this change is interesting to you, suggest what changes you would like to see in the PR for serious consideration of accepting it. Thx.

# Testing, Review

For convenience of *our* local development we added Makefile and Dockerfile as separate commit. (Also removed airbnb linter requirement because of `pnpm install` conflict error. See  and ignore separate labeled commit for that.)

For convenience of change preview changed `/docs/../ResponsiveTimePickers.tsx` to display separately in ampm and 24h formats.

1. check out this branch
2. `make sh`
3. (inside `make sh`) `pnpm install`
4. `make serve`  (serves docs on 3001)
5. visit http://localhost:3001/x/react-date-pickers/time-picker/
